### PR TITLE
Typeswitch utility

### PR DESCRIPTION
**Pull request by apmccartney**
Added a struct useful for several metaprogramming techniques

### CODE REVIEW

**Comment by apmccartney** [comment]
Note the build failures are due to the llvm apt repository issue

File: TypeSwitch/test/TypeSwitch.test.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        LOG(INFO) << "Test " << ++testNumber
+                  << ": [TypeSwitch] No Errors Expected";
+        is_same = std::is_same
+                  < void, utility::TypeSwitch< int, void,
+                                               ( 1 == 0 ) >::type >::value;
+        REQUIRE( true == is_same );
+
+
+        LOG(INFO) << "Test " << ++testNumber
+                  << ": [TypeSwitch] No Errors Expected";is_not_same =
```
Comment (jlconlin) [inline TypeSwitch/test/TypeSwitch.test.cpp]: There is a slight typo here. Please insert newline.

**Comment by jlconlin** [comment]
I'm not sure exactly what this pull request solves. It seems you can give two different types and switch between them? Why?

File: TypeSwitch/test/TypeSwitch.test.cpp (hunk 1 above)
Comment (apmccartney) [inline TypeSwitch/test/TypeSwitch.test.cpp]: good catch

**Comment by apmccartney** [comment]
Is there an issue with this pull?

**Comment by jlconlin** [comment]
Sorry, I thought I had asked a question about it.

**Pull request merged**